vm/devices/storage/storvsc_driver/src/lib.rs

```diff
@@ -0,0 +1,856 @@
+    UnsupportedProtocolVersion,
+    /// Unexpected protocol data or operation.
+    #[error("unexpected protocol data or operation")]
+    UnexpectedOperation,
+    /// Error notifying completion of operation.
+    #[error("error notifying completion of operation")]
+    NotificationError,
+    /// Error sending request to storvsc driver.
+    #[error("error sending request to storvsc")]
+    RequestError,
+    /// Error receiving new operations from channel.
+    #[error("error receiving new operations from channel")]
+    RequestReceiveError(#[source] RecvError),
+    /// Error waiting for completion of operation.
+    #[error("error waiting for completion of operation")]
+    CompletionError(#[source] RecvError),
+    /// Operation cancelled.
+    #[error("pending operation cancelled")]
+    Cancelled,
+    /// Storvsc driver not fully initialized.
+    #[error("driver not initialized")]
+    Uninitialized,
+}
+
+/// Errors with packet parsing between storvsc and storvsp.
+#[derive(Debug, Error)]
+pub enum PacketError {
+    /// Not transactional.
+    #[error("Not transactional")]
+    NotTransactional,
+    /// Unenxpected transaction.
+    #[error("Unexpected transaction {0:?}")]
+    UnexpectedTransaction(u64),
+    /// Unexpected status.
+    #[error("Unexpected status {0:?}")]
+    UnexpectedStatus(storvsp_protocol::NtStatus),
+    /// Unrecognzied operation.
+    #[error("Unrecognized operation {0:?}")]
+    UnrecognizedOperation(storvsp_protocol::Operation),
+    /// Invalid packet type.
+    #[error("Invalid packet type")]
+    InvalidPacketType,
+    /// Invalid data transfer length.
+    #[error("Invalid data transfer length")]
+    InvalidDataTransferLength,
+    /// Access error.
+    #[error("Access error: {0}")]
+    Access(#[source] AccessError),
+    /// Range error.
+    #[error("Range error")]
+    Range(#[source] ExternalDataError),
+}
+
+impl<T: 'static + Send + Sync + RingMem> StorvscDriver<T> {
+    /// Create a new driver instance connected to storvsp over VMBus.
+    pub fn new(
+        driver_source: &VmTaskDriverSource,
+        version: storvsp_protocol::ProtocolVersion,
+    ) -> Self {
+        Self {
+            storvsc: TaskControl::new(StorvscState),
+            version,
+            driver_source: driver_source.clone(),
+            new_request_sender: None,
+        }
+    }
+
+    /// Start Storvsc.
+    pub fn run(&mut self, channel: RawAsyncChannel<T>, target_vp: u32) -> Result<(), StorvscError> {
+        let driver = self
+            .driver_source
+            .builder()
+            .target_vp(target_vp)
+            .run_on_target(true)
+            .build("storvsc");
+        let (new_request_sender, new_request_receiver) =
+            async_channel::unbounded::<StorvscRequest>();
+        let storvsc = Storvsc::new(channel, self.version, new_request_receiver)?;
+        self.new_request_sender = Some(new_request_sender);
+
+        self.storvsc.insert(&driver, "storvsc", storvsc);
```

Comment:
I'd probably move negotiation here so that the caller can know whether the connection was successful. Otherwise, it may be hard to debug why SCSI requests are just stuck or whatever.